src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,632 @@
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(-EROFS);
+    return;
+  }
+
+  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
+      handle_update_object_map(r);
+      finish_op_ctx->complete(0);
+    });
+
+  RWLock::WLocker object_map_locker(m_dst_image_ctx->object_map_lock);
+  bool sent = m_dst_image_ctx->object_map->template aio_update<
+    Context, &Context::complete>(
+      snap_object_state.first, m_dst_object_number, snap_object_state.second,
+      {}, {}, ctx);
+  assert(sent);
+  m_dst_image_ctx->snap_lock.put_read();
+  m_dst_image_ctx->owner_lock.put_read();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::handle_update_object_map(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  assert(r == 0);
+  if (!m_snap_object_states.empty()) {
+    send_update_object_map();
+    return;
+  }
+  finish(0);
+}
+
+template <typename I>
+Context *ObjectCopyRequest<I>::start_dst_op(RWLock &owner_lock) {
+  assert(m_dst_image_ctx->owner_lock.is_locked());
+  if (m_dst_image_ctx->exclusive_lock == nullptr) {
+    RWLock::RLocker snap_locker(m_dst_image_ctx->snap_lock);
+    if (m_dst_image_ctx->test_features(RBD_FEATURE_EXCLUSIVE_LOCK,
+                                       m_dst_image_ctx->snap_lock)) {
+      return nullptr;
+    }
+    return new FunctionContext([](int r) {});
+  }
+  return m_dst_image_ctx->exclusive_lock->start_op();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::init_offsets() {
+  m_snap_copy_ops = {};
+  m_snap_object_states = {};
+  m_snap_object_sizes = {};
+  m_src_object_offsets = {};
+
+  std::vector<std::pair<uint64_t, uint64_t>> file_extents;
+  Striper::extent_to_file(m_cct, &m_dst_image_ctx->layout, m_dst_object_number,
+                          0, m_dst_image_ctx->layout.object_size, file_extents);
+
+  for (auto &e : file_extents) {
+    std::map<object_t, std::vector<ObjectExtent>> src_object_extents;
+    Striper::file_to_extents(m_cct, m_src_image_ctx->format_string,
+                             &m_src_image_ctx->layout, e.first, e.second, 0,
+                             src_object_extents);
+    for (auto &p : src_object_extents) {
+      for (auto &s : p.second) {
+        m_src_object_offsets[s.objectno].push_back({s.offset, s.length});
+      }
+    }
+  }
+
+  assert(!m_src_object_offsets.empty());
+
+  ldout(m_cct, 20) << m_src_object_offsets.size() << " src objects, first: "
+                   << *m_src_object_offsets.begin() << dendl;
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_diffs() {
+  auto iter = m_src_object_offsets.rbegin();
+  assert(iter != m_src_object_offsets.rend());
+  auto src_object_number = iter->first;
+  bool first_src_object = (m_src_object_offsets.size() == 1);
+
+  for (auto &it : iter->second) {
+    auto src_object_off = it.first;
+    auto len = it.second;
+
+    interval_set<uint64_t> copy_interval;
+    copy_interval.insert(src_object_off, src_object_off + len);
```

Comment:
Thinking about the complexity of this some more, I wonder if it would make more sense to refactor the ```CopyOp``` computation logic and separate out the read and write ops into two different collections. The reads would be ordered by write/read snapshots and source object and the writes could be ordered by write snapshot.

We would change the flow a little so that for each source object, we first list its diff then perform the read. This allows us to avoid repeating all source object reads should a retry be required. Once all the reads have been performed across all source objects, we can perform the write stage.

When computing the diffs, instead of creating remove/zero/truncate ops, we could instead just track an interval set of zeroed locations for the destination object at the current write snapshot. Then as part of the write step, we would generate zero, truncate or remove ops I think pretty easily.